### PR TITLE
feat: VSCODE-37: Marketplace setup and research

### PR DESCRIPTION
**Pull request by imlucas**
## Description

I'm walking through [the docs](https://code.visualstudio.com/api/working-with-extensions/publishing-extension) to see how this all actually works. 

- [x] Updates for marketplace meta @fredtruman posted below (HT @mmarcon)
- [x] Visual Studio Marketplace: Publisher `mongodb` (setup but waiting to hear back from MS support so we can add `vscode-marketplace@mongodb.com` as an owner)
- [x] Finish updates to `README.md` (move to another doc per max's comment below)
- [x] ~Windows: `##[warning]No test result files matching test-results.xml were found.`~ fix later
- [x] Fix triejs dep causing azure pipeline failures [COMPASS-4191](https://jira.mongodb.org/browse/COMPASS-4191)
- [x] fix: VSCODE-63
- [x] Setup email aliases
- [x] Setup [`vscode-marketplace@mongodb.com`](https://groups.google.com/a/mongodb.com/forum/#!forum/vscode-marketplace) to use as contact for Visual Studio Marketplace
- [x] Setup [`vscode@mongodb.com`](https://groups.google.com/a/mongodb.com/forum/#!forum/vscode) as contact email to use for this project
- [x] Low hanging fruit of [marketplace customization](https://code.visualstudio.com/api/references/extension-manifest#marketplace-presentation-tips) for @mmarcon to tweak (`homepage`, `gallery`, etc. in package.json)
- [x] `npm run package` script+step in azure pipeline to 1. make sure nothing unexpected 2. we can publish these artifacts as incrementals for debugging

### fix: VSCODE-63: syntax and snippets weren't included in .vsix 

The `.vscodeignore` `src/*` means the entire src directory not included in packaged bundle.  So made scripts write these assets to resources instead. Also, having some fun making scripts pretty :p

### Checklist
- [x] New tests and/or benchmarks are included
- [x] Documentation is changed or added

### CODE REVIEW

**Comment by imlucas** [comment]
Screenshot 2020-03-02 15 59 35

**Comment by imlucas** [comment]
### Future stuff for@mmarcon

- Read more carefully and update for any musts: https://code.visualstudio.com/api/references/extension-manifest#marketplace-presentation-tips
- Pattern for examples/libraries to [add mongodb-vscode as recommended extension](https://code.visualstudio.com/docs/editor/extension-gallery#_workspace-recommended-extensions)
- Finish design for driver specific bundling like `MongoDB for .Net`, `MongoDB for Mongoose`, etc. using [extension packs](https://code.visualstudio.com/api/references/extension-manifest#extension-packs)

#### metrics

- [sessionId and machineId from `vscode.env`](https://github.com/microsoft/vscode-extension-telemetry/blob/master/src/telemetryReporter.ts#L92)
- Extension install/uninstall hooks https://code.visualstudio.com/api/references/extension-manifest#extension-uninstall-hook

**Comment by imlucas** [comment]
Warning in tests:

```
No tree view with id 'mongoDB' registered.: Error: No tree view with id 'mongoDB' registered.
	at t.ExtHostTreeViews.$setVisible (/home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:518:566)
	at p._doInvokeHandler (/home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:676:202)
	at p._invokeHandler (/home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:675:901)
	at p._receiveRequest (/home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:674:545)
	at p._receiveOneMessage (/home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:673:373)
	at /home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:671:588
	at l.fire (/home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:48:541)
	at _.fire (/home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:190:934)
	at /home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:838:649
	at l.fire (/home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:48:541)
	at _.fire (/home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:190:934)
	at t.PersistentProtocol._receiveMessage (/home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:195:277)
	at /home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:192:459
	at l.fire (/home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:48:541)
	at p.acceptChunk (/home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:188:445)
	at /home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:187:796
	at Socket.t (/home/vsts/work/1/s/.vscode-test/vscode-1.42.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:196:732)
	at Socket.emit (events.js:200:13)
	at addChunk (_stream_readable.js:294:12)
	at readableAddChunk (_stream_readable.js:275:11)
	at Socket.Readable.push (_stream_readable.js:210:10)
	at Pipe.onStreamRead (internal/stream_base_commons.js:166:17)
```

File: package.json [hunk 1]
```diff
@@ -1,92 +1,101 @@
 {
```
Comment (mmarcon) [inline package.json]: @imlucas sorry for messing with the formatting of the file, actually VS Code did it automatically. We can revert the changes if it creates merge conflicts.

File: package.json [hunk 2]
```diff
@@ -1,92 +1,101 @@
...
-  "main": "./out/extension.js",
-  "repository": {
-    "type": "git",
-    "url": "https://github.com/mongodb-js/vscode"
+  "preview": true,
+  "icon": "resources/mongodb.png",
+  "galleryBanner": {
+    "color": "#21313D",
+    "theme": "dark"
   },
```
Comment (mmarcon) [inline package.json]: @fredtruman can you take a look at this stuff? Documentation here: https://code.visualstudio.com/api/references/extension-manifest#marketplace-presentation-tips

In short:
- `icon`: the icon that is displayed in the marketplace (the word count icon in [this example](https://marketplace.visualstudio.com/items?itemName=ms-vscode.wordcount)
- `color`: the color of the header in the marketplace (the red in [this example](https://marketplace.visualstudio.com/items?itemName=ms-vscode.wordcount))

Comment (fredtruman) [reply package.json]: Hey @mmarcon - recommending this:
Screenshot 2020-03-05 14 00 18

**Assets and color specs:**
Banner background:
#3D4F58

Button:
Static: #13AA52
Hover: #0B3B35

Icon for header:
mongodb-icon

Comment (mmarcon) [reply package.json]: @fredtruman I might be wrong but I don't think we can control the color of the button. Looks like it's the same for all the extensions.

Static: #107c10
Hover: #AAAAAA

Comment (fredtruman) [reply package.json]: Ah gotcha. Ok I think this gray background still works with the existing/standard button colors

File: package.json (hunk 1 above)
Comment (imlucas) [reply package.json]: No worries! I was just fidgeting :p `.vscode` auto-formatting on save is very intentional

**Comment by imlucas** [comment]
> `preview` (boolean) | Sets the extension to be flagged as a Preview in the Marketplace.

I set to `true` so we get the little badge.

> `qna` `marketplace` (default), `string`, `false` | Controls the Q & A link in the Marketplace. Set to marketplace to enable the default Marketplace Q & A site. Set to a string to provide the URL of a custom Q & A site. Set to false to disable Q & A altogether.

Added link to new community forums `https://community.mongodb.com/c/vscode/`. Maybe we actually create that category to link directly to?


> `keywords`   array | An array of keywords to make it easier to find the extension. These are included with other extension Tags on the Marketplace. This list is currently limited to 5 keywords.

`MongoDB` and `Database` so we show up in these 2 lists:
Screenshot: extensions tagged `Database`
Screenshot: extensions tagged `MongoDB`

**Comment by imlucas** [comment]
> Notes originally in README i'll find a Better home for at some point.

---

### Packaging

`vscode-mongodb` will be published under the `mongodb` Visual Studio Marketplace Publisher that has been setup under the existing `team-compass` Azure DevOps org we use for pipelines.

Visual Studio Marketplace

- Publisher `mongodb`
- [`vscode-marketplace@mongodb.com`](https://groups.google.com/a/mongodb.com/forum/#!forum/vscode-marketplace) is a member of the `team-compass` Azure DevOps Org
- [`vscode@mongodb.com`](https://groups.google.com/a/mongodb.com/forum/#!forum/vscode) is contact email to use for this project

Azure DevOps Org:
`team-compass`

[Azure Org Owner User](https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/change-organization-ownership?view=azure-devops):
`team-compass@mongodb.com`

Azure Projects:

- `vscode` The MongoDB VSCode Extension https://github.com/mongodb-js/vscode
- `mongosh` The MongoDB Shell https://github.com/mongodb-js/mongosh
- `compass` Compass plugins

* [docs](https://code.visualstudio.com/api/working-with-extensions/publishing-extension)
* [web ui](https://marketplace.visualstudio.com/manage/publishers/mongodb)
* Analytics: [Extension Reporting Hub](https://devblogs.microsoft.com/devops/extension-reporting-hub-for-marketplace-publishers/)

#### Preview Releases

While [private extensions are not currently supported](https://github.com/microsoft/vscode/issues/21839), the extension can be packaged as a `.vsix` file and shared directly.

To create a `.vsix`:

```bash
npm i -g vsce;
vsce package;
```

This will create a file like `mongodb-vscode-0.0.1.vsix`. Once downloaded, the extension can be installed with the [Install .VSIX](https://github.com/fabiospampinato/vscode-install-vsix) or by running:

```bash
code --install-extension mongodb-vscode-0.0.1.vsix
```

**Comment by imlucas** [comment]
@mmarcon updated for another peak. Could you open a new JIRA (or just draft a PR) for what you're thinking for screenshots/polish/adding preview release disclaimer etc. to the README? Everything we want to do just before we ship basically.
